### ISSUE TITLE
Referencing office-ui-fabric-react as external

### ISSUE THREAD

**Issue by SPParseError** [SharePoint/sp-dev-docs#1344]
#### Category
- [X] Question
- [ ] Typo
- [ ] Bug
- [ ] Additional article idea

#### Expected or Desired Behavior
As I have multiple webparts in my project, I would like to reference office-ui-fabric-react as external, so that it does not get bundled with each webpart.

If I simply add the reference in externals in config.json, i do have an external file generated, and my webparts files go from ~800kb to ~100kb. But when I test it in the workbench, or in a "production" site, I get an error saying the component has a dependency on a module "office-ui-fabric-react" that is not referenced in the manifest. Error goes away if I add a "globalName" in my external reference in config.json (procedure for non AMD modules). But then a javascript error occures because it gets loaded before React.

However, doing researches, I find [in the official documentation](https://docs.microsoft.com/en-us/sharepoint/dev/spfx/web-parts/get-started/use-fabric-react-components) that I should use Fabric React that shipped with the generator. Samples show non external references.

Scaffholded a new project with latest version of the generator. I take all references in package.json, update my project's package.json and reinstall everything in node_modules. I still need to add (npm install) office-ui-fabric-react in my project to be able to reference its components

My question is : 
1. Should I reference office-ui-fabric-react as external ? Is it still supported ?
2. Should I be able to reference office-ui-fabric-react without adding it with npm install ?

Thanks !

**Comment by nickpape-msft** [comment]
You were following this guidance: https://docs.microsoft.com/en-us/sharepoint/dev/spfx/web-parts/basics/add-an-external-library#share-a-library-among-multiple-web-parts ?

You should be able to reference office-ui-fabric-react as an external, but you must `npm install` it first. What will happen is that a separate chunk (JS file) will be creating containing all the office-ui-fabric-react code. This chunk will be shipped along with your web parts.

One thing to be aware of is that office-ui-fabric-react has breaking changes from time-to-time so it cannot be packaged in the same way as the SPFx Framework libraries, which are guaranteed to be backwards compatible.

**Comment by VesaJuvonen** [comment]
hi @SPParseError, since seems to be answered by our engineering, closing for now. Let's reopen if needed. Thx.

**Comment by SPParseError** [comment]
Hi, 

I tried it once again. 

Here is my externals section in config.json : 
![image](https://user-images.githubusercontent.com/12406819/36210756-8cfe160a-119f-11e8-9e05-f3ab7d8092b8.png)

Here is my import statement in the webpart : 
![image](https://user-images.githubusercontent.com/12406819/36210788-a6c19c42-119f-11e8-9890-0c3a20bcf34f.png)

I have this error when running gulp serve : 
![image](https://user-images.githubusercontent.com/12406819/36210706-71ab2604-119f-11e8-95f0-8f54787a4828.png)

Looks like I would need to add a "globalName" statement in the external declaration. If I do, I don't have this error. But then the code fires an error saying "React" does not exist, and I can't declare "React" as a dependency in externals.

@VesaJuvonen should we reopen it ?

**Comment by jonthenerd** [comment]
Seeing the same behavior @SPParseError . This makes it a particular pain to externalize any module that has a dependency upon React. The version of React that loads into the page (15.6.2 at this time) doesn't expose itself as a global variable, which office-ui-fabric is trying to access.

**Comment by giuleon** [comment]
Cancel

On 22 Mar 2018 20:37, "Jon Badgett" <notifications@github.com> wrote:

Seeing the same behavior @SPParseError <https://github.com/spparseerror> .
This makes it a particular pain to externalize any module that has a
dependency upon React. The version of React that loads into the page
(15.6.2 at this time) doesn't expose itself as a global variable, which
office-ui-fabric is trying to access.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/SharePoint/sp-dev-docs/issues/1344#issuecomment-375431830>,
or mute the thread

**Comment by harshdamaniahd** [comment]
Any solution?

**Comment by VesaJuvonen** [comment]
It is not unfortunately supported to reference Office UI Fabric as an external package as that will cause you version exceptions. Recommendation and how to use Office UI Fabric React is covered in following documentation - https://docs.microsoft.com/en-us/sharepoint/dev/spfx/office-ui-fabric-integration. If you have any issues with this, would suggest to open a new issue so that your comments or findings will not get buried on closed issues.

**Comment by nishantverma-nagarro** [comment]
The UI Fabric React is not was not working using the node_module link. So we got a workaround to fix this issue.
Add the CDN path and globalName of the to config.json 
"externals": {
    "office-ui-fabric-react": {
      "path": "https://unpkg.com/office-ui-fabric-react@6.143.0/dist/office-ui-fabric-react.min.js",
      "globalName": "Fabric"
    }
  }

This will compile properly. However if you deploy and add the web part to a page it will through react not found error.
To remove the react error I added the [CDN path](https://reactjs.org/docs/cdn-links.html) to the page using SEWP on classic page. In case of Modern page you can create extension to add the react js.

Once this is done the web part loads fine.

I know this is not the correct way but this will get your web parts working and will not create large bundle files.
